controllers/pulp/utils.go

```diff
@@ -245,3 +250,200 @@ func mountCASpec(pulp *repomanagerv1alpha1.Pulp, volumes []corev1.Volume, volume
 
 	return volumes, volumeMounts
 }
+
+// checkImmutableFields verifies if a user tried to modify an immutable field and rollback
+// the change if so
+func (r *PulpReconciler) checkImmutableFields(ctx context.Context, pulp *repomanagerv1alpha1.Pulp, field string, log logr.Logger) bool {
+
+	// access the field by its string name
+	// for fieldSpec we need to pass it as a reference because we will need to change
+	// its value back in case of immutable field
+	fieldSpec := reflect.Indirect(reflect.ValueOf(&pulp.Spec)).FieldByName(field)
+
+	// for fieldStatus, as we just need its content, we dont need to get a
+	// pointer to it
+	fieldStatus := reflect.ValueOf(pulp.Status).FieldByName(field)
+
+	// first we need to call the Interface() method to use the field as interface{}
+	// then we assert that the interface{} is a string so we can check if the len > 0
+	// if len>0 means that the field was previously defined
+	// after that we check if the content from .status.field is different from spec.field
+	// if so we should rollback the modification
+	if len(fieldStatus.Interface().(string)) > 0 && fieldStatus.Interface() != fieldSpec.Interface() {
+		patch := client.MergeFrom(pulp.DeepCopy())
+
+		// set pulp.spec.<field> back with the value from .status
+		fieldSpec.SetString(fieldStatus.Interface().(string))
+
+		// we are using patch here because we need to modify only a specific field.
+		// if we had used update it would fill a lot of other fields with default values
+		// which would also trigger a reconciliation loop
+		r.Patch(ctx, pulp, patch)
+		err := fmt.Errorf("%s field is immutable", field)
+		log.Error(err, "Could not update "+field+" field")
+		return true
+	}
+	return false
+}
+
+// updateIngressType will check the current definition of ingress_type and will handle the different
+// modification scenarios
+func (r *PulpReconciler) updateIngressType(ctx context.Context, pulp *repomanagerv1alpha1.Pulp) {
+
+	// if route => nodeport
+	// delete all routes with operator's labels
+	// remove route .status.conditions
+	// update .status.ingress_type = nodeport
+	if strings.ToLower(pulp.Status.IngressType) == "route" && strings.ToLower(pulp.Spec.IngressType) == "nodeport" {
+
+		route := &routev1.Route{}
+		routesLabelSelector := map[string]string{
+			"pulp_cr": pulp.Name,
+			"owner":   "pulp-dev",
+		}
+		listOpts := []client.DeleteAllOfOption{
+			client.InNamespace(pulp.Namespace),
+			client.MatchingLabels(routesLabelSelector),
+		}
+		r.DeleteAllOf(ctx, route, listOpts...)
```

Comment:
this wouldn't be repeated if it was under:
Status.IngressType == "route" and .Spec.IngressType != "route"

Reply:
Oh ... I see what you mean now.
You are assuming that for all of the scenarios
route => ingress
route => nodeport
route => loadbalancer
the behavior will be the same, so we don't need to handle each one of them and could just check if it was defined as route and now the user is defining `.spec.ingress_type` to anything else.

Nice! I'm updating it.